components/brave_federated/config_utils.cc

```diff
@@ -0,0 +1,130 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_federated/config_utils.h"
+
+#include "base/files/file_util.h"
+#include "base/json/json_reader.h"
+#include "base/strings/string_number_conversions.h"
+#include "brave/components/brave_federated/features.h"
+
+namespace brave_federated {
+
+LearningServiceConfig::LearningServiceConfig() {
+  reconnect_policy_ = {.num_errors_to_ignore = 0,
+                       .initial_delay_ms = 10 * 1000,
+                       .multiply_factor = 2.0,
+                       .jitter_factor = 0.0,
+                       .maximum_backoff_ms = 60 * 10 * 1000,
+                       .always_use_initial_delay = true};
+  request_task_policy_ = {
+      .num_errors_to_ignore = 0,
+      .initial_delay_ms =
+          features::GetFederatedLearningUpdateCycleInSeconds() * 1000,
+      .multiply_factor = 2.0,
+      .jitter_factor = 0.0,
+      .maximum_backoff_ms =
+          16 * features::GetFederatedLearningUpdateCycleInSeconds() * 1000,
+      .always_use_initial_delay = true};
+  post_results_policy_ = {
+      .num_errors_to_ignore = 0,
+      .initial_delay_ms =
+          features::GetFederatedLearningUpdateCycleInSeconds() * 1000,
+      .multiply_factor = 2.0,
+      .jitter_factor = 0.0,
+      .maximum_backoff_ms =
+          16 * features::GetFederatedLearningUpdateCycleInSeconds() * 1000,
+      .always_use_initial_delay = true};
+  model_spec_.num_params = 32, model_spec_.batch_size = 32,
+  model_spec_.learning_rate = 0.01, model_spec_.num_iterations = 500,
+  model_spec_.threshold = 0.5;
+}
+
+LearningServiceConfig::LearningServiceConfig(const base::FilePath& path)
+    : LearningServiceConfig::LearningServiceConfig() {
+  std::string data;
+  const bool success = base::ReadFileToString(path, &data);
+  if (!success || data.empty()) {
+    VLOG(2) << "FL: Error in reading JSON configuration from " << path;
+    return;  // return default ctor's initialisation
```

Comment:
See https://abseil.io/tips/42 as seems strange to fail and return in the constructor

Reply:
cc @stevelaskaridis

Reply:
I am not sure I understand the point here. There are default values in the default ctor that are overridden in the presence of the JSON configuration. The "failure" is an internal object state that is never communicated to the outside world.

In the presence of a factory method, it would simply handle the very same if..then condition in a separate function. It would not return null as indicated by the url.

Essentially:
```
if valid(json_configuration):
  return LearningServiceConfig(json_configuration)
else:
  return LearningServiceConfig(default_configuration)  # it does not fail here
```

Unless you are suggesting that this constructor chaining is problematic, which I am not sure why ...

Reply:
I missed the default constructor... and the reason was that it is normally good practice if a constructor cannot construct a valid object, to throw an exception. However, in Chromium we should not throw exceptions so for me I fall back to the above when there is a possibility of failure